### PR TITLE
util: fix some encoding issue when setting unicode/utf8 titles in py2

### LINKED ISSUE #2444: UnicodeDecodeError when opening streams with non-ascii characters in the title
## Bug Report

- [x] This is a bug report and I have read the contribution guidelines.

### Description

When opening a stream with a non-ascii character in the stream title there is a “UnicodeDecodeError” and the stream doesn’t open.

The steam I tried to open had an ‘ö’ in the title, but trying other streams with non-ascii characters gave the same error (tested on random Korean streams). Other streams opened as expected.

### Expected / Actual behavior

Expect: Stream to open.
Actual: It doesn’t.

### Reproduction steps / Explicit stream URLs to test

1. Find a stream with a non-ascii title and attempt to open it.

### Log output

```
		OS: macOS 10.14.2
		Python: 2.7.14
		Streamlink: 1.1.1
		Requests(2.21.0), Socks(1.6.7), Websocket(0.47.0)
		Found matching plugin twitch for URL twitch.tv/quickybaby
		Plugin specific arguments:
		--twitch-oauth-token=******** (oauth_token)
		--twitch-disable-hosting=True (disable_hosting)
		--twitch-disable-ads=True (disable_ads)
		Getting live HLS streams for quickybaby
		Attempting to authenticate using OAuth token
		Successfully logged in as DELETE
		[utils.l10n][debug] Language code: en_US
		Available streams: audio_only, 160p (worst), 360p, 480p, 720p, 720p60, 1080p60 (best)
		Opening stream: 1080p60 (hls)
		Starting player: /Applications/mpv.app/Contents/MacOS/mpv
		Traceback (most recent call last):
		File "/usr/local/bin/streamlink", line 11, in <module>
		load_entry_point('streamlink==1.1.1', 'console_scripts', 'streamlink')()
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/main.py", line 1033, in main
		handle_url()
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/main.py", line 594, in handle_url
		handle_stream(plugin, streams, stream_name)
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/main.py", line 437, in handle_stream
		success = output_stream_passthrough(plugin, stream)
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/main.py", line 262, in output_stream_passthrough
		output.open()
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/output.py", line 24, in open
		self._open()
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/output.py", line 219, in _open
		self._open_call()
		File "/usr/local/lib/python2.7/site-packages/streamlink-1.1.1-py2.7.egg/streamlink_cli/output.py", line 234, in _open_call
		log.debug(u"Calling: {0}".format(fargs))
		UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 141: ordinal not in range(128)
```

### Additional comments, screenshots, etc.
I can't link to a stream that will give an error as titles always change, but it should be easy enough to find a Korean stream, and they usually have Korean titles.

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

### PR DESCRIPTION
**Pull request by beardypig**
This should resolve the encoding issue in log messages for the command line arguments. 

fixes #2444

### CODE REVIEW

**Comment by back-to** [comment]
there is also a similar bug since https://github.com/streamlink/streamlink/pull/2765


```
streamlink https://www.twitch.tv/videos/584715362 best --title "{title}"

[cli][debug] OS:         Linux
[cli][debug] Python:     2.7.15+

Traceback (most recent call last):
  File "py2/bin/streamlink", line 11, in <module>
    load_entry_point('streamlink', 'console_scripts', 'streamlink')()
  File "src/streamlink_cli/main.py", line 1052, in main
    handle_url()
  File "src/streamlink_cli/main.py", line 604, in handle_url
    handle_stream(plugin, streams, stream_name)
  File "src/streamlink_cli/main.py", line 457, in handle_stream
    success = output_stream(plugin, stream)
  File "src/streamlink_cli/main.py", line 321, in output_stream
    output.open()
  File "src/streamlink_cli/output.py", line 24, in open
    self._open()
  File "src/streamlink_cli/output.py", line 223, in _open
    self._open_subprocess()
  File "src/streamlink_cli/output.py", line 244, in _open_subprocess
    args = self._create_arguments()
  File "src/streamlink_cli/output.py", line 194, in _create_arguments
    extra_args.append("--title={}".format(self.title))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-1: ordinal not in range(128)
```

---

won't work on **windows**

```
C:\streamlink-2444-fix-unicode-title>C:\Python27\Scripts\streamlink.exe -l debug twitch.tv/twitch worst --title "ä"
[cli][debug] OS:         Windows 10
[cli][debug] Python:     2.7.16
[cli][debug] Streamlink: 2444-fix-unicode-title
[cli][debug] Requests(2.22.0), Socks(1.7.0), Websocket(0.56.0)
[cli][info] Found matching plugin twitch for URL twitch.tv/twitch
[plugin.twitch][debug] Getting live HLS streams for twitch
[cli][info] Available streams: audio_only, 160p (worst), 360p, 480p, 720p, 720p60, 1080p60 (best)
[cli][info] Opening stream: 160p (hls)
[stream.hls][debug] Reloading playlist
[stream.hls][debug] First Sequence: 7990; Last Sequence: 8003
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 8001; End Sequence: None
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 8001 to queue
[stream.hls][debug] Adding segment 8002 to queue
[stream.hls][debug] Adding segment 8003 to queue
[stream.hls][debug] Download of segment 8001 complete
[stream.hls][debug] Download of segment 8002 complete
[stream.hls][debug] Download of segment 8003 complete
[cli][info] Starting player: "C:\Program Files\VideoLAN\VLC\vlc.exe"
[cli][info] Closing currently open stream...
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
Traceback (most recent call last):
  File "C:\Python27\Scripts\streamlink-script.py", line 11, in <module>
    load_entry_point('streamlink', 'console_scripts', 'streamlink')()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 1024, in main
    handle_url()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 602, in handle_url
    handle_stream(plugin, streams, stream_name)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 455, in handle_stream
    success = output_stream(plugin, stream)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 319, in output_stream
    output.open()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\output.py", line 24, in open
    self._open()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\output.py", line 223, in _open
    self._open_subprocess()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\output.py", line 244, in _open_subprocess
    args = self._create_arguments()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\output.py", line 210, in _create_arguments
    eargs = maybe_decode(subprocess.list2cmdline(extra_args))
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink\utils\encoding.py", line 28, in maybe_decode
    return text.decode(encoding)
  File "C:\Python27\lib\encodings\utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 21: unexpected end of data
```

**Comment by beardypig** [comment]
@back-to that was easy enough to fix on linux/macOS, but testing on Windows is a bit painful for me. Could you help me out and see if the latest commit fixes it on Windows?

File: src/streamlink_cli/output.py [hunk 1]
```diff
@@ -91,6 +91,9 @@ def __init__(self, cmd, args=DEFAULT_PLAYER_ARGUMENTS, filename=None, quiet=True
         self.namedpipe = namedpipe
         self.http = http
         self.title = title
+        if self.title is not None:
+            # straiten up the encoding
+            self.title = maybe_decode(self.title)
```
Comment (back-to) [inline src/streamlink_cli/output.py:96]:
```
C:\Python27\Scripts\streamlink.exe -l debug twitch.tv/twitch worst --title "ä"
[cli][debug] OS:         Windows 10
[cli][debug] Python:     2.7.16
[cli][debug] Streamlink: 2444-fix-unicode-title
...
[stream.hls][debug] Adding segment 6897 to queue
[stream.hls][debug] Download of segment 6895 complete
[cli][info] Starting player: "C:\Program Files\VideoLAN\VLC\vlc.exe"
[cli][info] Closing currently open stream...
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
Traceback (most recent call last):
  File "C:\Python27\Scripts\streamlink-script.py", line 11, in <module>
    load_entry_point('streamlink', 'console_scripts', 'streamlink')()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 1024, in main
    handle_url()
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 602, in handle_url
    handle_stream(plugin, streams, stream_name)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 455, in handle_stream
    success = output_stream(plugin, stream)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 316, in output_stream
    output = create_output(plugin)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\main.py", line 125, in create_output
    record=record, title=title)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink_cli\output.py", line 96, in __init__
    self.title = maybe_decode(self.title)
  File "c:\streamlink-2444-fix-unicode-title\src\streamlink\utils\encoding.py", line 28, in maybe_decode
    return text.decode(encoding)
  File "c:\python27\lib\encodings\utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xe4 in position 0: unexpected end of data
```

Comment (beardypig) [reply src/streamlink_cli/output.py:96]: Thanks. I think I will need to set up a Windows VM to debug this :)

**Comment by beardypig** [comment]
This latest version worked for me.
```
streamlink https://www.twitch.tv/videos/584715362 worst --title "{title} ä"
```

I tested on
  - Python 2.7, Linux Ubuntu 18.04
  - Python 3.6, Linux Ubuntu 18.04
  - Python 2.7, macOS Catalina
  - Python 3.7, macOS Catalina
  - Python 2.7, Windows 10 Pro

I didn't get Python 3.7 working on Windows 10 Pro yet. Missing DLLs or something :/

**Comment by gravyboat** [comment]
Good stuff, thanks.

**Comment by beardypig** [comment]
2.7 is bad at unicode, when do we drop 2.7 support? :)

**Comment by gravyboat** [comment]
@beardypig ASAP!

**Pull request merged**